### PR TITLE
perf(encoder): donor-shape Fast kernel modules (#198 phase 1a)

### PR DESCRIPTION
**Pull request by polaz**
## Summary

Phase 1a of the Fast strategy donor port (issue #198 — 22× regression
vs C zstd on Fast levels). This PR lands the donor-shape kernel
modules with full unit-test coverage; the wiring into
`MatchGeneratorDriver` / `MatcherStorage` and the i9-9900K bench land
in the follow-up PR on the next branch.

## What's in this PR

Three new modules under `zstd/src/encoding/simple/fast_kernel/`:

- **`hash_table.rs` — `FastHashTable`**
  Flat `Vec<u32>` indexed by donor-parity `ZSTD_hash{4,5,6,7,8}Ptr`
  (multiply-shift on the first `mls` bytes of the suffix at `ptr`,
  output reduced to `hash_log` bits). Constants bit-identical to
  `lib/compress/zstd_compress_internal.h`. Monomorphised over `MLS`
  via const-generic so each `mls` instantiation compiles to a
  dedicated body with the prime / shift baked in.

- **`count.rs` — `count_forward`**
  Port of `ZSTD_count` from `lib/compress/zstd_compress_internal.h`.
  8-byte chunked XOR loop with `trailing_zeros() / 8` to locate the
  diverging byte, then `u32` / `u16` / `u8` tail fall-throughs
  matching donor's exact progression. Replaces the byte-by-byte
  `common_prefix_len` used by the current SuffixStore-based Fast
  matcher.

- **`kernel.rs` — `compress_block_fast<MLS>`**
  Port of `ZSTD_compressBlock_fast_noDict_generic` from
  `lib/compress/zstd_fast.c`. Phase 1 keeps the single-`ip0` cursor
  with raw 4-byte match probe (`MEM_read32(ip) == MEM_read32(base +
  matchIdx)`), step-based skip acceleration tied to
  `kSearchStrength = 6`, repcode-at-ip check, backward extension
  bounded by `anchor` / `prefix_start`, and `count_forward` for
  forward extension. The 4-cursor pipelining (`ip0 / ip1 / ip2 /
  ip3`) and `cmov` match-found variant from donor's full generic
  body are deferred to phase 3 per the issue's staged plan — phase 1
  exists to validate the data structures and close the bulk of the
  22× gap before adding pipelining complexity.

The kernel's `compress_block_fast` is monomorphised over `const MLS:
u32` (4..=8), invokes `handle_sequence` per emitted sequence
(replacing donor's `ZSTD_storeSeq`), and returns
`FastBlockResult { rep, tail_literals_len }` so the caller can
thread repcode history across blocks and flush trailing literals.

## What's NOT in this PR (lands in follow-up)

- Wiring into `MatchGeneratorDriver::compress_block::<Fast>` — phase
  1b. Requires:
  - Adding `FastHashTable` field to `MatchGenerator` (alongside the
    existing `SuffixStore` during transition).
  - Flat-history-buffer materialisation (concat `Vec<WindowEntry>`
    into a single `Vec<u8>` per `start_matching` call, or replace
    the multi-window layout with a single ring buffer).
  - Hash-table alloc / clear in `reset` keyed off the resolved
    `LevelParams` `(hash_log, mls)` — donor uses `hash_log = 14`,
    `mls = 7` for default level 1.
- Encoder/decoder roundtrip validation (current `cargo nextest run
  --workspace` covers existing SuffixStore path; the kernel hasn't
  been wired yet so it can't break the e2e suite from this PR).
- i9-9900K bench on `compare_ffi level_1_fast` — measures the actual
  perf delta vs main and against C donor. Phase 1b commit will land
  the numbers in its PR body.

Phase 3 (separate later branch per the issue):

- 4-cursor `ip0/ip1/ip2/ip3` pipelining with software-prefetch hash
  reads — gives the bulk of donor's noDict_generic throughput on
  modern superscalar cores.
- `cmov` match-found variant gated on `windowLog < 19`.
- Cross-block matches via prefix history (phase 1a's kernel is
  single-block; cross-block matching at the window boundary is part
  of phase 1b's flat-buffer work).

## Tests

16 new unit tests covering:

- `hash_table::tests`: donor formula parity on known inputs for both
  `mls=4` and `mls=5`, get/put round-trip, `clear()` resets every
  entry to sentinel.
- `count::tests`: empty / full / chunk-boundary / tail-fall-through
  divergence at every transition point (u32 → u16 → u8), thousand-
  byte long match.
- `kernel::tests`: short-input → literals-only callback path, long
  repeat pattern → at least one `Triple` emitted with `match_len ≥
  MIN_MATCH (4)`, plus an accounting invariant
  (`sum(literals + matches) + tail_literals_len == input.len()`)
  that catches off-by-ones in any future refactor.

All 16 pass; full workspace nextest run remains green at 555/555.

## Module gate

`fast_kernel/mod.rs` opens with a narrowly-scoped
`#![allow(dead_code)]` (without `unused_imports`) for this commit
only — every item is consumed by the phase 1b wiring commit and
the allow is removed there. The `unused_imports` lint stays ACTIVE
across the module so any accidentally-unused `use` in sibling code
surfaces immediately. Re-exports of the kernel's public surface
(`FastHashTable`, `FastBlockResult`, `compress_block_fast`) are
deferred to the wiring commit since adding them now would force
the wider lint relaxation. Without the `dead_code` allow, baseline
`cargo clippy -D warnings` would flag the unused kernel-internal
items; splitting the kernel and the wiring into separate PRs is
the right tradeoff for review clarity, but the cost is the
temporary `dead_code` mute on this module.

Related: #198 (Fast strategy 22× regression — multi-phase port).

## Summary by CodeRabbit

* **New Features**
  * Added a donor-style "Fast" block compressor to the simple encoder for improved match finding, rep/offset handling, and faster compression.
  * Added an optimized forward match counter and a compact power-of-two hash table to improve match-extension accuracy and throughput.

* **Tests**
  * Added unit tests covering matching behavior, hash-table operations, compressor outcomes, and edge cases (empty inputs, early mismatches, chunk/tail boundaries, repcode behavior, long full-match).

[Review Change Stack](https://app.coderabbit.ai/change-stack/structured-world/structured-zstd/pull/215?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: zstd/src/encoding/simple/fast_kernel/kernel.rs [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+        if tail_literals_len > 0 {
+            handle_sequence(Sequence::Literals {
+                literals: &data[block_start..],
+            });
+        }
+        return FastBlockResult {
+            rep,
+            tail_literals_len: 0,
+        };
+    }
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/simple/fast_kernel/kernel.rs:287]: @coderabbitai Fixed — short-input early-return path now also defers tail emission to the caller (returns tail_literals_len = data.len() - block_start without emitting Literals). Function doc updated with the explicit sequence-emission contract that applies uniformly to every exit branch.

File: zstd/src/encoding/simple/fast_kernel/hash_table.rs [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    pub(crate) fn new(hash_log: u32, mls: u32) -> Self {
+        assert!(
+            hash_log > 0,
+            "hash_log must be > 0 (a zero log would make hash_ptr shift by the full word width)",
+        );
+        assert!(
+            hash_log < usize::BITS,
+            "hash_log {hash_log} >= usize::BITS {} would overflow `1usize << hash_log`",
+            usize::BITS,
+        );
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/simple/fast_kernel/hash_table.rs]: @coderabbitai Fixed — tightened to 1..=ZSTD_HASHLOG_MAX (30), matching donor's documented cap. Both invalid bands (lower-bound shift overflow + mls=4 32-bit-shift overflow) now rejected at construction.

File: zstd/src/encoding/simple/fast_kernel/hash_table.rs [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+    /// doesn't reject 30 via a panic-catching probe.
+    #[test]
+    fn hash_log_maximum_thirty_is_accepted_by_constructor() {
+        // Just probe the assertion path: invoking new() with 30
+        // mustn't panic. We don't actually use the giant table — drop
+        // it immediately. (2^30 * 4 bytes is acceptable for a CI host
+        // running a single test; if the host can't allocate it the
+        // test will OOM-skip rather than misreport.)
+        let _table = FastHashTable::new(30, 4);
+    }
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/simple/fast_kernel/hash_table.rs:410]: @coderabbitai Fixed — extracted validate_params() helper, test now calls it directly with ZSTD_HASHLOG_MAX (and mls 4/8) without any allocation. FastHashTable::new delegates to the same helper so panic behaviour is unchanged.

File: zstd/src/encoding/simple/fast_kernel/kernel.rs [hunk 4]
```diff
@@ -0,0 +1,799 @@
...
+            //
+            // Combined, every loop iteration that enters this branch
+            // already satisfies `new_ip - 1 - rep_off >=
+            // prefix_start`, so the extra check would be dead code on
+            // the hot path. If a future call shape weakens either
+            // invariant (e.g. shared hash table across resets without
+            // re-running save/restore) the explicit guard must be
+            // re-added — see the explicit-match path at the
+            // `match_pos > prefix_start_index` line for the
+            // structurally symmetric bound that proves it.
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/simple/fast_kernel/kernel.rs]: @coderabbitai Fixed — rewrote the comment to make the three-piece argument explicit: (1) non-strict bound from save/restore, (2) strict bound from explicit-match promotion (covers iterations after first promote), (3) runtime `new_ip > anchor` gate (covers the block-entry boundary iteration where (1) alone gives equality and (2) hasn't kicked in yet). Added an explicit note that weakening any of the three — including anchor initialisation — requires the explicit prefix guard.

File: zstd/src/encoding/simple/fast_kernel/kernel.rs [hunk 5]
```diff
@@ -0,0 +1,875 @@
...
+/// The `ilimit = data.len() - HASH_READ_SIZE` cap constrains where
+/// the main loop hashes and probes — i.e. it stops emitting new
+/// matches once `ip0 >= ilimit`. It does NOT mean the trailing 7
+/// bytes are ALWAYS literals: an in-progress forward match found at
+/// `ip0 < ilimit` extends through `count_forward` and can reach all
+/// the way to `iend`, leaving `tail_literals_len = 0`. The kernel
+/// reports the actual number of trailing literal bytes (zero or
+/// more) in `FastBlockResult.tail_literals_len`, and the caller
+/// emits a terminal `Sequence::Literals` only when that value is
+/// non-zero.
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/simple/fast_kernel/kernel.rs]: @coderabbitai Fixed — reworded the precondition to require `data.len() - block_start >= HASH_READ_SIZE` (the remaining-block length), and added a pointer to the matching short-input guard. The new wording is consistent with the runtime check at the line right below.

**Comment by polaz** [comment]
@coderabbitai Fixed — replaced `debug_assert_eq!(MLS, hash_table.mls())` with two real `assert!`s in `compress_block_fast`: a range check `(4..=8).contains(&MLS)` and a strict `assert_eq!` matching MLS against the table's mls. Both run in every build now, so a release-mode `compress_block_fast::<5>(..., FastHashTable::new(_, 4), ...)` panics with a clear diagnostic instead of silently miscompressing.

**Comment by polaz** [comment]
@coderabbitai Re: outside-diff MLS contract finding at kernel.rs:188-220 — this is STALE, already fixed in commit 183c36f3 (round 10). The kernel now does `assert!((4..=8).contains(&MLS))` + `assert_eq!(MLS, hash_table.mls())` instead of `debug_assert_eq!`. Both run in every build and panic with diagnostic messages on mismatch.

File: zstd/src/encoding/simple/fast_kernel/hash_table.rs [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+    /// - `entries * size_of::<u32>()` overflowing `usize` (same
+    ///   class of overflow at a different multiply step).
+    /// - Global allocator failure when actually allocating the
+    ///   table backing storage — propagates as the standard
+    ///   `Vec::with_capacity` allocation-failure panic.
+    ///
+    /// The parameter-range and target-size guards both fire BEFORE
+    /// the `vec![]` call, so allocator failure (the third bullet)
+    /// is the only panic that depends on runtime memory state. The
+    /// first two are deterministic given the inputs and target.
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/simple/fast_kernel/hash_table.rs]: @coderabbitai Fixed — see commit 12e67517.

File: zstd/src/encoding/simple/fast_kernel/kernel.rs [hunk 7]
```diff
@@ -0,0 +1,924 @@
...
+/// into the short-input early-return branch and emits no sequences,
+/// which may not be what the caller wanted.
+///
+/// # Panics
+///
+/// Entry-time `assert!`s reject misuse loudly in every build (debug
+/// AND release) rather than silently miscompressing:
+/// - `block_start > data.len()` — would wrap
+///   `block_start + HASH_READ_SIZE` in the short-input guard and
+///   skip the early return.
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/simple/fast_kernel/kernel.rs]: @coderabbitai Fixed — see commit 12e67517.

File: zstd/src/encoding/simple/fast_kernel/hash_table.rs [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+    ///   surfaces this as a clear panic instead of the opaque
+    ///   `Vec`-internal capacity-overflow message.
+    /// - Global allocator failure when actually allocating the
+    ///   table backing storage — propagates as the standard
+    ///   `Vec::with_capacity` allocation-failure panic.
+    ///
+    /// All three guards fire BEFORE control reaches `vec![]`, so
+    /// allocator failure (the third bullet) is the only panic that
+    /// depends on runtime memory state. The first two are
+    /// deterministic given the inputs and target architecture.
```
[the remark this answers is not in the document]
Comment (polaz) [reply zstd/src/encoding/simple/fast_kernel/hash_table.rs]: @coderabbitai Fixed in aa3faf5e — reworded to say only the first two guards run pre-vec; the third bullet IS the vec![] allocation itself (the only runtime-state-dependent panic).
